corehq/util/couch.py

```diff
@@ -33,3 +36,130 @@ def categorize_bulk_save_errors(error):
         result_map[error].append(result)
 
     return result_map
+
+
+class IterDB(object):
+    """
+    Bulk save docs in chunks.
+
+        with IterDB(db) as iter_db:
+            for doc in iter_docs(db, ids):
+                iter_db.save(doc)
+
+        iter_db.error_ids  # docs that errored
+        iter_db.saved_ids  # docs that saved correctly
+    """
+    def __init__(self, database, chunksize=100, throttle_secs=None):
+        self.db = database
+        self.chunksize = chunksize
+        self.throttle_secs = throttle_secs
+        self.saved_ids = []
+        self.deleted_ids = []
+        self.error_ids = []
+
+    def __enter__(self):
+        self.to_save = []
+        self.to_delete = []
+        return self
+
+    def commit(self, cmd, docs):
+        try:
+            results = cmd(docs)
+        except BulkSaveError as e:
+            results = e.results
+            error_ids = {d['id'] for d in e.errors}
```

Comment:
There is, and Simon's got a nice function splitting it up.  I've used that to provide an attribute [`errors_by_type`](https://github.com/dimagi/commcare-hq/commit/1669045bdb038c163ae754b9ed28f5931eba9280#diff-4f793748a49396e1612026f0d9e999eeL68)